semantics/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+//! Fe semantic analysis.
+//!
+//! This library is used to analyze the semantics of a given Fe AST. It detects
+//! any semantic errors within a given AST and produces a `Context` instance
+//! that can be used to query contextual information attributed to AST nodes.
+
+pub mod errors;
+pub mod namespace;
+mod traversal;
+
+use crate::errors::SemanticError;
+use crate::namespace::events::Event;
+use crate::namespace::scopes::Shared;
+use crate::namespace::types::{
+    FixedSize,
+    Type,
+};
+use fe_parser::ast as fe;
+use fe_parser::span::{
+    Span,
+    Spanned,
+};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+/// Indicates where an expression is stored.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Location {
+    Storage { index: usize },
+    Memory,
+    Value,
+}
+
+/// Contains contextual information relating to an expression AST node.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ExpressionAttributes {
+    pub typ: Type,
+    pub location: Location,
+}
+
+/// Contains contextual information relating to a function definition AST node.
+#[derive(Clone, Debug, PartialEq)]
+pub struct FunctionAttributes {
+    pub name: String,
+    pub param_types: Vec<FixedSize>,
+    pub return_type: Option<FixedSize>,
+}
+
+/// Contains contextual information about a Fe module and can be queried using
+/// `Spanned` AST nodes.
+#[derive(Clone, Debug, Default)]
+pub struct Context {
+    expressions: HashMap<Span, ExpressionAttributes>,
+    emits: HashMap<Span, Event>,
+    functions: HashMap<Span, FunctionAttributes>,
+    declarations: HashMap<Span, FixedSize>,
+    contracts: HashMap<Span, Vec<FunctionAttributes>>,
```

Comment:
`contracts` holding a `HashMap<Span, Vec<FunctionAttributes>>` makes it look as if a contract would only hold functions but it can also hold declarations and events.

https://github.com/ethereum/fe/blob/503d7bb9c2d4241e6a6b38982756ae875714d775/compiler/tests/fixtures/guest_book.fe#L4-L7

I'm wondering if instead of `FunctionAttributes` a type should be used that would combine `FunctionAttributes`, `Event`  and `FixedSize`?

Reply:
Oh, I just finished reading the whole PR description and noticed:

>The API of context was built out only to the extent needed by this PR

>f multiple things need to be attributed to one type of AST node, they should be grouped together using an attributes struct e.g. `ContractAttributes`

I *think* that this is basically saying what I wrote above but waiting for you to confirm. If yes, than I think a small comment above line `58` would make a nice improvement in the meantime.

Reply:
Yeah, I can add a comment here.